File: hws/attrs.py
from attrs import *
import sys


def has_attr(attrs, x):
    if x in attrs:
        return True
    else:
        return False

def remove_attr(attrs, x):
    dict1 = dict(attrs)
    for key in list(dict1.keys()):
        if key == x:
            dict1.pop(key)
    print(dict1)

def update_attr(attrs, x):
    dict1 = dict(attrs)
    for key in list(dict1.keys()):
        if key == x[0] and x[1] > 0:
            dict1[key] = x[1]
            print(dict1)
            return
        elif key == x[0] and x[1] < 0:
            dict1.pop(key)
            print(dict1)
            return
    dict1[x[0]] = x[1]
    print(dict1)

def update_multi_attr(attrs, x):
    y = x[1]
    x = x[0]
    xstate = True
    ystate = True

    dict1 = dict(attrs)
    while xstate == True:
        if x[1] < 0:
            xstate = False
        for key in list(dict1.keys()):
            if key == x[0] and x[1] >= 0:
                dict1[key] = x[1]
                xstate = False
            elif key == x[0] and x[1] < 0:
                dict1.pop(key)
                xstate = False

    while ystate == True:
        if y[1] < 0:
            ystate = False
        for key in list(dict1.keys()):
            if key == y[0] and y[1] >= 0:
                dict1[key] = y[1]
                ystate = False
            elif key == y[0] and y[1] < 0:
                dict1.pop(key)
                ystate = False

    print(dict1)

non_attr_val = -sys.maxsize

def attrs_from_string(x):
    dict1 = {}
    x = x.split(" ")
    for i in range(0,len(x),2):
        dict1[x[i]] = int(x[i+1])
    print(dict1)
    
            

def main():
    attrs = {'a': 3, 'b': 4}
    has_attr(attrs, "b")
    remove_attr(attrs, 'a')
    update_attr(attrs, ['a', 17])
    update_attr(attrs, ['c', 17])
    print(non_attr_val)
    update_attr(attrs, ['a', non_attr_val])
    update_multi_attr(attrs, [['b', 5], ['a', non_attr_val]])
    attrs_from_string("d 3 e 4")


if __name__ == "__main__":
    main()

File: hw8/attrs/attrs.go
//package attrs

package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

type Attr struct {
	x string
	y int
}

func HasAttr(x []Attr, y string) bool {
	for _, value := range x {
		if value.x == y {
			return true
		}
		//fmt.Print(value)
	}
	return false
}

func RemoveAttr(x []Attr, y string) []Attr {
	var count int
	for index, value := range x {
		if value.x == y {
			count = index
		}
	}
	copylist := append(x[:count], x[count+1:]...)
	return copylist
}

var NonAttrVal = math.MinInt32

func UpdateAttr(x []Attr, y Attr) []Attr {
	var count int = 0
	var attrs []Attr
	for _, value := range x {
		if y.y == math.MinInt32 {
			//x[count] = nil
			attrs = append(x[:count], x[count+1:]...)
			//x = j
			//fmt.Println(j)
			//fmt.Print(attrs)
			return attrs
		} else if value.x == y.x {
			x[count].y = y.y
		}
		count++
	}
	//fmt.Println(x)
	//fmt.Print(x)
	return x
}

func UpdateMultiAttr(x []Attr, y []Attr) []Attr {
	var count int = 0
	for _, value := range x {
		for _, value2 := range y {
			if value.x == value2.x {
				//k := &x[count]
				//k.y = value2.y
				x[count].y = value2.y
			}
		}
		count++
	}
	for i := 0; i < len(x); i++ {
		if x[0].y == math.MinInt32 {
			x = append(x[:i], x[i+1:]...)
			i--
		}
	}
	//fmt.Println(x)
	//fmt.Print(x)
	return x
}

func AttrsFromString(x string) []Attr {
	var attrs []Attr
	s := strings.Split(x, " ")
	//fmt.Println(len(s))
	for i := 0; i < len(s); i = i + 2 {
		num, _ := strconv.Atoi(s[i+1])
		//fmt.Println(num, num2)
		attrs = append(attrs, Attr{s[i], num})
	}
	return attrs
}

func main() {
	attrs := []Attr{{"a", 3}, {"b", 4}}
	//fmt.Println(attrs)
	//y := RemoveAttr(attrs, "a")
	//fmt.Println(y)
	//fmt.Println(NonAttrVal)
	//fmt.Println(attrs)
	x := UpdateAttr(attrs, Attr{"a", 17})
	fmt.Println(x)
	//fmt.Println(attrs)
	//UpdateAttr(attrs, Attr{"a", NonAttrVal})
	//fmt.Println(attrs)
	//UpdateMultiAttr(attrs, []Attr{{"b", 5}, {"a", NonAttrVal}})
	fmt.Println(attrs)
	//AttrsFromString("d 3 e 4")
	//fmt.Println(attrs)
}
